Make conform respect config.vim.lsp.formatOnSave and config.vim.lsp.mappings.toggleFormatOnSave

venkyr77 opened the pull request:
Update of PR https://github.com/NotAShelf/nvf/pull/764

The current defaults([link1](https://github.com/alfarelcynthesis/nvf/blob/main/modules/plugins/formatter/conform-nvim/conform-nvim.nix#L35-L38), [link2](https://github.com/alfarelcynthesis/nvf/blob/main/modules/plugins/formatter/conform-nvim/conform-nvim.nix#L47))from conform config is not respecting the top level config we have for `formatOnSave` - [vim.lsp.formatOnSave](https://notashelf.github.io/nvf/options.html#opt-vim.lsp.formatOnSave). Also, `config.vim.lsp.mappings.toggleFormatOnSave` has no impact.

From [conform docs](https://github.com/stevearc/conform.nvim?tab=readme-ov-file#options)

```
-- If this is set, Conform will run the formatter on save.
-- It will pass the table to conform.format().
-- This can also be a function that returns the table
```

```
-- If this is set, Conform will run the formatter asynchronously after save.
-- It will pass the table to conform.format().
-- This can also be a function that returns the table.
```

these need to be set to `null` (or `nil` in `lua`) when `config.vim.lsp.formatOnSave` is `false`

This PR fixes conform config to respect `config.vim.lsp.formatOnSave`

## Testing:

Tested my changes by changing the `nvf` url to the branch from my fork

```
nvf.url = "github:venkyr77/nvf/conform-format-on-save-fix";
```

with `config.vim.lsp.formatOnSave = false;` conform is not auto formatting the file

with `config.vim.lsp.formatOnSave = true;`

![formatOnSaveTrue](https://github.com/user-attachments/assets/d5868549-44ef-425b-ad70-28db1d932f3b)

with `config.vim.lsp.formatOnSave = true;` and using the mapping for `config.vim.lsp.mappings.toggleFormatOnSave`

![toggleFormatOnSave](https://github.com/user-attachments/assets/258e700a-2959-4759-952b-bbc5dae8a311)

<!--
^ Please include a clear and concise description of the aim of your Pull Request above this line ^

For plugin dependency/module additions, please make sure to link the source link of the added plugin
or dependency in this section.

If your pull request aims to fix an open issue or a please bug, please also link the relevant issue
below this line. You may attach an issue to your pull request with `Fixes #<issue number>` outside
this comment, and it will be closed when your pull request is merged.

A developer package template is provided in flake/develop.nix. If working on a module, you may use
it to test your changes with minimal dependency changes.
-->

## Sanity Checking

<!--
Please check all that apply. As before, this section is not a hard requirement but checklists with more checked
items are likely to be merged faster. You may save some time in maintainer reviews by performing self-reviews
here before submitting your pull request.

If your pull request includes any change or unexpected behaviour not covered below, please do make sure to include
it above in your description.
-->

[editorconfig]: https://editorconfig.org
[changelog]: https://github.com/NotAShelf/nvf/tree/main/docs/release-notes
[hacking nvf]: https://notashelf.github.io/nvf/index.xhtml#sec-guidelines

- [ ] I have updated the [changelog] as per my changes
- [x] I have tested, and self-reviewed my code
- [x] My changes fit guidelines found in [hacking nvf]
- Style and consistency
  - [x] I ran **Alejandra** to format my code (`nix fmt`)
  - [x] My code conforms to the [editorconfig] configuration of the project
  - [x] My changes are consistent with the rest of the codebase
- If new changes are particularly complex:
  - [ ] My code includes comments in particularly complex areas
  - [ ] I have added a section in the manual
  - [ ] _(For breaking changes)_ I have included a migration guide
- Package(s) built:
  - [x] `.#nix` _(default package)_
  - [x] `.#maximal`
  - [x] `.#docs-html` _(manual, must build)_
  - [ ] `.#docs-linkcheck` _(optional, please build if adding links)_
- Tested on platform(s)
  - [x] `x86_64-linux`
  - [ ] `aarch64-linux`
  - [ ] `x86_64-darwin`
  - [ ] `aarch64-darwin`

<!--
If your changes touch upon a portion of the codebase that you do not understand well, please make sure to consult
the maintainers on your changes. In most cases, making an issue before creating your PR will help you avoid duplicate
efforts in the long run. `git blame` might help you find out who is the "author" or the "maintainer" of a current
module by showing who worked on it the most.
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

venkyr77 commented:
@horriblename, fixes https://github.com/NotAShelf/nvf/pull/764#issuecomment-2765974470

venkyr77 commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,68 @@ in {
+              function()
+                if (not vim.g.formatsave) or (vim.b.disableFormatSave) then
+                  return
+                else
+                  return ${toLuaObject defaultFormatOnSaveOpts}
```
nit: no else return

if the changes look good, I can make a fast followup

horriblename commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,68 @@ in {
+            # lua
+            ''
+              function()
+                if (not vim.g.formatsave) or (vim.b.disableFormatSave) then
```
1. useless parentheses
2. in conform docs, they used `vim.b[bufnr].disableFormatSave` where  bufnr is the first argument of the function

venkyr77 replied:
> in conform docs, they used vim.b[bufnr].disableFormatSave where bufnr is the first argument of the function

Both works, I think this way fits easily into the current autocmd way of setting `vim.g.formatsave` and `vim.b.disableFormatSave`. I drew inspiration from - https://github.com/stevearc/conform.nvim/issues/39#issuecomment-1937061763

horriblename replied:
fair, but remove useless parentheses please

venkyr77 replied:
done, removed

horriblename commented on modules/plugins/lsp/config.nix:
```diff
@@ -93,11 +93,7 @@ in {
-                bufnr = bufnr,
-              })
-        "
+          else ""
```
don't remove this completely, we still want it if conform and null-ls are both disabled

venkyr77 replied:
> don't remove this completely, we still want it if conform and null-ls are both disabled

added the change back

horriblename commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,68 @@ in {
-          save.
-        '';
-      };
+            Note:
+              - When config.vim.lsp.formatOnSave is set to true, internally
+                vim.g.formatsave is set to true.
+              - vim.b.disableFormatSave initally equals !config.vim.lsp.formatOnSave.
+              - vim.b.disableFormatSave is toggled using the
+                mapping from config.vim.lsp.mappings.toggleFormatOnSave.
```
`vim.g.formatsave` and `vim.b.disableFormatSave` are internal stuff and not meant to be part of any API. (unrelated: I'd give them more consistent names if we were to expose them)

remove this section

venkyr77 replied:
> remove this section

removed

horriblename reviewed (commented):
Looks good other than some documentation stuff

NotAShelf commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,68 @@ in {
+          type = nullOr (either attrs luaInline);
+          default =
+            mkLuaInline
+            # lua
```
```suggestion
```

Please don't do this. I understand it's easier to handle the code with syntax highlighting, but we don't inject syntax anywhere else in the source tree and Alejandra results in horrible formatting when you specific a language.

Easier if we provide injections for mkLuaInline, CC @horriblename

horriblename replied:
#781

horriblename replied:
fwiw, I find flaw in both approaches. Ours require custom query rules for each option we want to support which is tedious, and theirs enforce a style of coding that is usually not in line with official language guidelines just to support a feature in a text editor plugin not everyone uses

venkyr77 replied:
>Please don't do this

removed

NotAShelf commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,68 @@ in {
+              end
+            '';
+          description = ''
+            Table or function(lualinline) that will be passed to `conform.format()`. If this
```
```suggestion
            Table or function(lualinline) that will be passed to `conform.format()`. If this
```

Is this a function that takes `luainline` as an argument?

venkyr77 replied:
>  Is this a function that takes luainline as an argument?

it can take both

a table or a function that returns table

NotAShelf replied:
Okay, I'm mostly confused about the role of the paranthesis in the description. Does it indicate function arguments? 

I think it's better to clarify the wording to state the types properly.


```suggestion
            Attribute set or or a Lua function that will be passed to `conform.format()`.
```

Or similar.

NotAShelf commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,68 @@ in {
-          save.
-        '';
-      };
+            Note:
+              - When config.vim.lsp.formatOnSave is set to true, internally
+                vim.g.formatsave is set to true.
+              - vim.b.disableFormatSave initally equals !config.vim.lsp.formatOnSave.
+              - vim.b.disableFormatSave is toggled using the
+                mapping from config.vim.lsp.mappings.toggleFormatOnSave.
```
```suggestion
            :::{.note}
              - When config.vim.lsp.formatOnSave is set to true, internally
                vim.g.formatsave is set to true.
              - vim.b.disableFormatSave initally equals !config.vim.lsp.formatOnSave.
              - vim.b.disableFormatSave is toggled using the
                mapping from config.vim.lsp.mappings.toggleFormatOnSave.
					  :::
```

Nixpkgs commonmark can handle note styles. Also put vim globals etc inside backticks.

venkyr77 replied:
>Nixpkgs commonmark can handle note styles. Also put vim globals etc inside backticks.

I just removed the section as @horriblename suggested removing the section as these are just internal APIs

NotAShelf commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,68 @@ in {
+          type = nullOr (either attrs luaInline);
+          default =
+            mkLuaInline
+            # lua
```
```suggestion
```

venkyr77 replied:
fixed

NotAShelf commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,68 @@ in {
+            Table or function(luainline) that will be passed to `conform.format()`. If this
+            is set, Conform will run the formatter asynchronously after save.
+
+            Note:
+              - When config.vim.lsp.formatOnSave is set to true, internally
+                vim.g.formatsave is set to true.
+              - vim.b.disableFormatSave initally equals !config.vim.lsp.formatOnSave.
+              - vim.b.disableFormatSave is toggled using the
+                mapping from config.vim.lsp.mappings.toggleFormatOnSave.
```
Use note markup as described above.

venkyr77 replied:
I just removed the section as @horriblename suggested removing the section as these are just internal APIs

horriblename commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,68 @@ in {
+      in
+        mkOption {
+          type = nullOr (either attrs luaInline);
+          default =
```
also: use lib.options.literalMD in the defaultText

venkyr77 commented:
Diff addressing all the comments - https://github.com/NotAShelf/nvf/compare/7eec1d28d2a8f510964f304fb460884ef9f48b4c..0ffd8321e9fee37e3f191209921546a40523a269

venkyr77 commented:
> Diff addressing all the comments - https://github.com/NotAShelf/nvf/compare/7eec1d28d2a8f510964f304fb460884ef9f48b4c..0ffd8321e9fee37e3f191209921546a40523a269

@horriblename @NotAShelf

NotAShelf commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,48 @@ in {
+            end
+          '';
+          description = ''
+            Table or function(luainline) that will be passed to `conform.format()`. If this
+            is set, Conform will run the formatter asynchronously after save.
```
Same as my previous comment.

NotAShelf reviewed (changes requested):
Two small notes on the documentation. I have nothing further to nit (for once)

Perhaps a changelog entry would also go very well with the spirit of this change?

horriblename commented on modules/plugins/formatter/conform-nvim/conform-nvim.nix:
```diff
@@ -30,27 +27,48 @@ in {
+      in
+        mkOption {
+          type = nullOr (either attrs luaInline);
+          default = mkLuaInline ''
```
add defaultText please

something like

```
defaultText = literalMD "enabled, with fallback to LSP formatting that respects `vim.lsp.formatOnSave` and `vim.lsp.mappingd.toggleFormatSave`" 
```

horriblename commented:
superseded by #891
